### PR TITLE
G migration mechanism

### PR DESCRIPTION
**Pull request by alex-courtis**
Migrate old g: options to setup. Much better user experience. I was surprised by window_picker migration ;)

`nvim-tree-startup.lua` -> `legacy.lua`

g: will be migrated if it is not present in the options passed to setup.

All migrations tested extensively.

### CODE REVIEW

File: doc/nvim-tree-lua.txt [hunk 1]
```diff
@@ -613,7 +613,7 @@ INFORMATIONS				        *nvim-tree-info*
 - <C-x> will open the file in a horizontal split
 - <C-t> will open the file in a new tab
 - <Tab> will open the file as a preview (keeps the cursor in the tree)
-- `I` will toggle visibility of folders hidden via |g:nvim_tree_ignore|
+- `I` will toggle visibility of folders hidden via |git.ignore| option
```
Comment (alex-courtis) [inline doc/nvim-tree-lua.txt:616]: typo

File: lua/nvim-tree.lua [hunk 2]
```diff
@@ -348,7 +349,7 @@ local DEFAULT_OPTS = {
   },
   filters = {
     dotfiles = false,
-    custom_filter = {},
+    custom = {},
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:352]: typo

File: lua/nvim-tree.lua [hunk 3]
```diff
@@ -364,9 +365,7 @@ local DEFAULT_OPTS = {
     open_file = {
       quit_on_open = vim.g.nvim_tree_quit_on_open == 1,
       window_picker = {
-        enable = vim.g.nvim_tree_disable_window_picker ~= 1,
-        chars = vim.g.nvim_tree_window_picker_chars,
-        exclude = vim.g.nvim_tree_window_picker_exclude,
+        enable = true,
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:368]: These were moved to legacy.lua

File: lua/nvim-tree/legacy.lua [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  end,
+
+  nvim_tree_auto_resize = function(o)
+    utils.table_create_missing(o, "view")
+    if o.view.auto_resize == nil then
+      o.view.auto_resize = vim.g.nvim_tree_auto_resize ~= 0
+    end
+  end,
+
+  nvim_tree_bindings = function(o)
```
Comment (alex-courtis) [inline lua/nvim-tree/legacy.lua:89]: This will work as long as tree_cb still exists.

File: lua/nvim-tree/legacy.lua [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+  nvim_tree_follow_update_path = function(o)
+    utils.table_create_missing(o, "update_focused_file")
+    if o.update_focused_file.update_cwd == nil then
+      o.update_focused_file.update_cwd = vim.g.nvim_tree_follow_update_path ~= 0
+    end
+  end,
+
+  nvim_tree_lsp_diagnostics = function(_)
+    -- option no longer exists
```
Comment (kyazdani42) [inline lua/nvim-tree/legacy.lua:79]: this is `diagnostics.enable` i think

**Review by kyazdani42 — APPROVED** [review]
very nice addition thanks. Could you just complete with the diagnostic entry ?

File: lua/nvim-tree/legacy.lua (hunk 5 above)
Comment (alex-courtis) [reply lua/nvim-tree/legacy.lua]: Thank you. Added `diagnostics.enable` and `diagnostics.show_on_dirs`.

Did not test actual diagnostics, just that the opts were successfully set/ignored.

TIL diagnostics and will set them up for :make

**Comment by alex-courtis** [comment]
Also... please update https://git.io/JPhyt to include

* `nvim_tree_disable_window_picker` -> `actions.open_file.window_picker.enable`
* `nvim_tree_window_picker_chars` -> `actions.open_file.window_picker.chars`
* `nvim_tree_window_picker_exclude` -> `actions.open_file.window_picker.exclude`

File: lua/nvim-tree/utils.lua [hunk 6]
```diff
@@ -225,4 +225,60 @@ function M.canonical_path(path)
...
+  for s in string.gmatch(sub, "([^%.]+)%.*") do
+    if t[s] == nil then
+      t[s] = {}
+    end
+    t = t[s]
+  end
+
+  return t
+end
+
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:247]: I have sneaked this in, even though this is not needed for the functionality. It is just useful.

Please feel free to direct me to remove it.

Comment (kyazdani42) [reply lua/nvim-tree/utils.lua:247]: isn't this something like `vim.inspect` ? I'll leave it here, remove it if you find `vim.inspect` to be a proper replacement :)

**Review by kyazdani42 — APPROVED** [review]

**Pull request merged**